File: auth-center/auth-center-domain/src/main/java/com/jmsoftware/maf/authcenter/user/service/impl/UserRoleDomainServiceImpl.java
package com.jmsoftware.maf.authcenter.user.service.impl;

import com.baomidou.mybatisplus.core.toolkit.Wrappers;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import com.jmsoftware.maf.authcenter.role.persistence.Role;
import com.jmsoftware.maf.authcenter.role.service.RoleDomainService;
import com.jmsoftware.maf.authcenter.user.mapper.UserRoleMapper;
import com.jmsoftware.maf.authcenter.user.persistence.User;
import com.jmsoftware.maf.authcenter.user.persistence.UserRole;
import com.jmsoftware.maf.authcenter.user.service.UserRoleDomainService;
import com.jmsoftware.maf.common.exception.BizException;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.stereotype.Service;

import javax.validation.constraints.NotBlank;
import java.util.Optional;

/**
 * UserRoleDomainServiceImpl.
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, date: 6/29/2021 12:09 PM
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class UserRoleDomainServiceImpl
        extends ServiceImpl<UserRoleMapper, UserRole>
        implements UserRoleDomainService {
    private final RoleDomainService roleDomainService;

    @Override
    @SneakyThrows({BizException.class})
    public void assignRoleByRoleName(@NonNull User user, @NotBlank String roleName) {
        val queryWrapper = Wrappers.lambdaQuery(Role.class);
        queryWrapper.select(Role::getId)
                .eq(Role::getName, roleName);
        val role = Optional.ofNullable(this.roleDomainService.getOne(queryWrapper))
                .orElseThrow(() -> new BizException("Cannot find the role: " + roleName));
        val userRole = new UserRole();
        userRole.setUserId(user.getId());
        userRole.setRoleId(role.getId());
        val saved = this.save(userRole);
        if (!saved) {
            throw new BizException(String.format("Cannot assign role (%s) to user (%s)", roleName, user.getUsername()));
        }
    }
}

File: auth-center/auth-center-domain/src/main/java/com/jmsoftware/maf/authcenter/user/service/impl/UserRoleDomainServiceImpl.kt
package com.jmsoftware.maf.authcenter.user.service.impl

import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import com.jmsoftware.maf.authcenter.role.persistence.Role
import com.jmsoftware.maf.authcenter.role.service.RoleDomainService
import com.jmsoftware.maf.authcenter.user.mapper.UserRoleMapper
import com.jmsoftware.maf.authcenter.user.persistence.User
import com.jmsoftware.maf.authcenter.user.persistence.UserRole
import com.jmsoftware.maf.authcenter.user.service.UserRoleDomainService
import com.jmsoftware.maf.common.exception.InternalServerException
import com.jmsoftware.maf.springcloudstarter.function.requireTrue
import org.springframework.stereotype.Service
import java.util.*

/**
 * # UserRoleDomainServiceImplKt
 *
 * Change description here.
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, 4/12/22 10:28 AM
 **/
@Service
class UserRoleDomainServiceImpl(
    private val roleDomainService: RoleDomainService
) : ServiceImpl<UserRoleMapper, UserRole>(), UserRoleDomainService {
    override fun assignRoleByRoleName(user: User, roleName: String) {
        val role = Optional.ofNullable(
            roleDomainService.ktQuery()
                .select(Role::id)
                .eq(Role::name, roleName)
                .one()
        ).orElseThrow { InternalServerException("Cannot find the role: $roleName") }
        val userRole = UserRole()
        userRole.userId = user.id
        userRole.roleId = role.id
        requireTrue(save(userRole), null).orElseThrow {
            InternalServerException("Cannot assign role ($roleName) to user (${user.username})")
        }
    }
}
